### DISCUSSION TITLE
pcp_combine SUBTRACT wrapper issues and general questions/comments

### DISCUSSION THREAD

**Discussion by mpm-meto** [dtcenter/METplus#1078] [Use Cases] [answered]
ERROR  : 
ERROR  : yyerror() -> syntax error in file "/tmp/met_config_25285_0"
ERROR  : 
ERROR  :    line   = 2
ERROR  : 
ERROR  :    column = 26
ERROR  : 
ERROR  :    text   = "(nul)"
ERROR  : 
ERROR  : 
ERROR  : (nul)
ERROR  : (nul)
ERROR  : 

I've been chasing this around for a bit... and couldn't work out what was wrong. Maybe this error message can be improved? It is a syntax error of sorts but the message does not point one to the fact that the field specification is incomplete, and that you need two parts: name and level in quotes e.g. -field 'name="stratiform_rainfall_amount"; level="(*,*)";' 

This error has arisen whilst trying to debug a pcp_combine wrapper for subtraction... I kept getting a level error
ERROR  : 
ERROR  : Dictionary::lookup_string() -> lookup failed for name "level"
ERROR  : 
even though the following are specified
FCST_PCP_COMBINE_INPUT_LEVELS = "(*,*)"
FCST_PCP_COMBINE_INPUT_NAME = stratiform_rainfall_amount

Running pcp_combine off the command line (without the wrapper) then gave me the yyerror() because I had not added the level string. I had just specified the name part as -field "stratiform_rainfall_amount".

Be that as it may, we can't understand why the wrapper is not taking the field information provided above and giving the lookup error, i.e. giving the level error even though it's specified and accepted by pcp_combine. 

First the wrapper does not appear to take note of the SUBTRACT in the wrapper and then proceeds to ignore the field info too.
**In bold are items NOT TRUE.** (unless there is another way to specifying this to setting FCST_PCP_COMBINE_INPUT_* .

**DEBUG 2: Since the "-field" command line option was not used**, parsing the command line arguments a list of files, each followed by a configuration string.
DEBUG 2: Performing **derivation command (sum)** for 1 files.
DEBUG 1: **Reading data (name="APCP"; level="A01"**;) from input file: /scratch/frmm/KSCALE/RA2T_2p2km/20171231T0000Z_KSCALE1_km2p2_ra2t_pverb001.nc
ERROR  : 
ERROR  : get_field() -> can't get data plane from file "/scratch/frmm/KSCALE/RA2T_2p2km/20171231T0000Z_KSCALE1_km2p2_ra2t_pverb001.nc"
ERROR  : 
08/11 15:04:00.443 metplus (command_builder.py:255) ERROR: (command_builder.py:1309) MET command returned a non-zero return code:/data/users/cfrd/MET_requirements/MET10.0.0/bin/pcp_combine -v 2 -add /scratch/frmm/KSCALE/RA2T_2p2km/20171231T0000Z_KSCALE1_km2p2_ra2t_pverb001.nc 01 /scratch/frmm/KSCALE/RA2T_2p2km/hourly/2017123100_F001_A01.nc
08/11 15:04:00.443 metplus INFO: Check the logfile for more information on why it failed: /data/users/frmm/METplus_Out/output/logs/metplus.log.20210811150359
08/11 15:04:00.443 metplus INFO: Processing forecast lead 2 hours
08/11 15:04:00.445 metplus INFO: ENVIRONMENT FOR NEXT COMMAND: 
08/11 15:04:00.445 metplus INFO: MET_TMP_DIR=/data/users/frmm/METplus_Out/output/tmp
08/11 15:04:00.446 metplus INFO: COMMAND: /data/users/cfrd/MET_requirements/MET10.0.0/bin/pcp_combine -v 2 -subtract /scratch/frmm/KSCALE/RA2T_2p2km/20171231T0000Z_KSCALE1_km2p2_ra2t_pverb002.nc 'name="stratiform_rainfall_amount";' /scratch/frmm/KSCALE/RA2T_2p2km/20171231T0000Z_KSCALE1_km2p2_ra2t_pverb001.nc 'name="stratiform_rainfall_amount";' /scratch/frmm/KSCALE/RA2T_2p2km/hourly/2017123100_F002_A01.nc
OUTPUT:
**DEBUG 2: Since the "-field" command line option was not used,** parsing the command line arguments a list of files, each followed by a configuration string.
DEBUG 1: Reading input file: /scratch/frmm/KSCALE/RA2T_2p2km/20171231T0000Z_KSCALE1_km2p2_ra2t_pverb002.nc
DEBUG 1: Reading data (name="stratiform_rainfall_amount";) from input file: /scratch/frmm/KSCALE/RA2T_2p2km/20171231T0000Z_KSCALE1_km2p2_ra2t_pverb002.nc
ERROR  : 
ERROR  : Dictionary::lookup_string() -> lookup failed for name "level"
ERROR  :

Finally, even though on the face of it pcp_combine without the wrapper ran through and produced an output file, we will check that pcp_combine SUBTRACT is working correctly because: 
(a) the model forecast files are accumulating from the initialisation time. 
(b) time_bnds in each file are set such that the first time_bnd value is always the forecast init time. I am not sure what pcp_combine will make of that. It doesn't appear to be doing things correctly. Essentially t+2 must be subtracted from t+1, then t+3 from t+2 etc. to get hourly totals irrespective of what the first time_bnd value is. 

ncdump` -t -v time_bnds /scratch/frmm/KSCALE/RA2T_2p2km/20171231T0000Z_KSCALE1_km2p2_ra2t_pverb001.nc
time_bnds = "2017-12-31", "2017-12-31 01" ;
ncdump` -t -v time_bnds /scratch/frmm/KSCALE/RA2T_2p2km/20171231T0000Z_KSCALE1_km2p2_ra2t_pverb002.nc
time_bnds = "2017-12-31", "2017-12-31 02" ;
ncdump` -t -v time_bnds /scratch/frmm/KSCALE/RA2T_2p2km/20171231T0000Z_KSCALE1_km2p2_ra2t_pverb003.nc
time_bnds = "2017-12-31", "2017-12-31 03" ;
etc

One other we notice is that the output from pcp_combine has the variable name with a _01 appended. I hope that can be overriden somehow because this is undesirable. The long name stays the same, but we wouldn't want to have the CF compliant name changing... am I right in thinking that whilst the netcdf input is CF compliant, the output netcdf from pcp_combine is not?

**Comment by mpm-meto** [comment]
For whatever reason the editor removed my (*,*) in FCST_PCP_COMBINE_INPUT_LEVELS = "(,)"!

**Comment by georgemccabe** [comment, marked as answer]
Hi @mpm-meto,

> I've been chasing this around for a bit... and couldn't work out what was wrong. Maybe this error message can be improved?

I agree that the error message from parsing the configuration is confusing. I would have to defer to @JohnHalleyGotway to see if there is a good way to improve the error message in this case. From what I understand these types of errors can occur in various places in the config parsing, so it may be difficult to advise as to what exactly has gone wrong here.

> First the wrapper does not appear to take note of the SUBTRACT in the wrapper and then proceeds to ignore the field info too.

From the log excerpt you provided, it looks like pcp_combine is being run twice, once using -add mode and once using -subtract mode. Could you also send the metplus_final.conf file that is generated in your OUTPUT_BASE directory so I can see what settings are being used for the run? If FCST_PCP_COMBINE_RUN = True and OBS_PCP_COMBINE_RUN = True, then the wrapper will run the tool twice for each run time.

Regarding the call to pcp_combine in subtract mode, the level value is not currently passed into the command for that mode. There is a note in the wrapper to add support for this, although we have not yet had a use case that needed it. With the work for issue #1062 I was planning on cleaning up the wrapper so that the different run methods are handled more consistently. I will be sure to enable setting the level value in the way you have shown when I clean up the PCPCombine wrapper.

I was under the impression that if the level value is "(\*,\*)" then you do not need to specify it explicitly (FYI, I put a backslash \ in front of the \* characters so they are not interpreted as a block of bold text). Could you send me the input files so I can take a closer look?

> One other we notice is that the output from pcp_combine has the variable name with a _01 appended. I hope that can be overriden somehow because this is undesirable.

The name of the output filename is controlled by [FCST/OBS]_PCP_COMBINE_OUTPUT_TEMPLATE, so the level (aka accumulation) is likely specified in the value for that variable. Adjusting the value will remove the extra info from the filename.

> The long name stays the same, but we wouldn't want to have the CF compliant name changing... am I right in thinking that whilst the netcdf input is CF compliant, the output netcdf from pcp_combine is not?

I'm not sure, but I didn't think that the name of the file is a requirement of CF compliance. However, you are correct to think that the output NetCDF file generated from pcp_combine is not CF compliant. The files generated by all of the MET tools are in their own format that is not currently CF compliant. There has been discussion to make the output CF compliant, but I am not sure how big of a task that would be. @JohnHalleyGotway may have more info on that as well.

I added a note in issue #1062 to ensure that the level can be passed into -subtract mode. If you'd like to send the files I mentioned in this comment I can take a closer look at the use case.

**Reply by JohnHalleyGotway** [reply]
@mpm-meto, the "name" and "level" strings are always required in those field strings. It is not the case that "(\*,\*)" is a default level string. However, I suppose we could add that as a default if needed.

While switching from the MET NetCDF format to a CF compliant version is desirable, it would require a significant amount of work. It is not a quick and easy change, and is not realistic to do without identifying time and funding ahead of time for it. In particular, we'd need to create library code for writing grids in a CF-compliant way. So definitely doable, but we'd need to plan for it.
